Source/JavaScriptCore/runtime/MicrotaskQueue.cpp

```diff
@@ -37,22 +41,106 @@ namespace JSC {
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(MicrotaskQueue);
 
-void QueuedTask::run()
+QueuedTask::Status QueuedTask::tryRun()
 {
+    if (m_dispatcher)
+        return m_dispatcher->run(*this);
+
     if (!m_job.isObject())
-        return;
+        return Status::Executed;
+
     JSObject* job = jsCast<JSObject*>(m_job);
-    JSGlobalObject* globalObject = job->globalObject();
-    runJSMicrotask(globalObject, m_identifier, job, m_arguments[0], m_arguments[1], m_arguments[2], m_arguments[3]);
+    runJSMicrotask(m_globalObject, m_identifier, job, m_arguments[0], m_arguments[1], m_arguments[2], m_arguments[3]);
+    return Status::Executed;
+}
+
+bool QueuedTask::isRunnable()
+{
+    if (!m_dispatcher)
+        return true;
+    return m_dispatcher->isRunnable();
 }
 
 MicrotaskQueue::MicrotaskQueue(VM& vm)
 {
     vm.m_microtaskQueues.append(this);
 }
 
+MicrotaskQueue::~MicrotaskQueue()
+{
+    if (isOnList())
+        remove();
+}
+
 template<typename Visitor>
 void MicrotaskQueue::visitAggregateImpl(Visitor& visitor)
+{
+    m_queue.visitAggregate(visitor);
+    m_toKeep.visitAggregate(visitor);
+}
+DEFINE_VISIT_AGGREGATE(MicrotaskQueue);
+
+void MicrotaskQueue::enqueue(QueuedTask&& task)
+{
+    auto* globalObject = task.globalObject();
+    auto microtaskIdentifier = task.identifier();
+    m_queue.enqueue(WTFMove(task));
+    if (globalObject) {
+        if (auto* debugger = globalObject->debugger(); UNLIKELY(debugger))
+            debugger->didQueueMicrotask(globalObject, microtaskIdentifier);
+    }
+}
+
+void MicrotaskQueue::performMicrotaskCheckpoint(VM& vm)
+{
+    auto catchScope = DECLARE_CATCH_SCOPE(vm);
+    while (!m_queue.isEmpty()) {
+        if (UNLIKELY(vm.executionForbidden())) {
+            clear();
+            break;
+        }
+
+        auto task = m_queue.dequeue();
+        auto status = task.tryRun();
+        if (UNLIKELY(!catchScope.clearExceptionExceptTermination())) {
+            clear();
+            break;
+        }
+
+        vm.callOnEachMicrotaskTick();
+        if (UNLIKELY(!catchScope.clearExceptionExceptTermination())) {
+            clear();
+            break;
+        }
+
+        switch (status) {
+        case QueuedTask::Status::Executed: {
+            break;
+        }
+        case QueuedTask::Status::Stopped: {
+            // Let this task go away.
+            break;
+        }
+        case QueuedTask::Status::Suspended: {
+            m_toKeep.enqueue(WTFMove(task));
+            break;
+        }
+        }
+    }
+    m_queue.swap(m_toKeep);
+}
+
+bool MarkedMicrotaskDeque::hasMicrotasksForFullyActiveDocument()
+{
+    for (auto& task : m_queue) {
+        if (task.isRunnable())
+            return true;
```

Comment:
It seems a bit inefficient to traverse through all these tasks to find the runnable one although in most cases, all tasks are runnable so we should exit in a couple of iterations. It's probably okay for this PR though.